node/cn/handler_test.go

```diff
@@ -58,3 +62,169 @@ func TestNewProtocolManager(t *testing.T) {
 		assert.Equal(t, errIncompatibleConfig, err)
 	}
 }
+
+func TestSampleSize(t *testing.T) {
+	peers := make([]Peer, minNumPeersToSendBlock-1)
+	assert.Equal(t, len(peers), sampleSize(peers))
+
+	peers = make([]Peer, 4)
+	assert.Equal(t, minNumPeersToSendBlock, sampleSize(peers))
+
+	peers = make([]Peer, 16)
+	assert.Equal(t, 4, sampleSize(peers))
+}
+
+func TestSamplePeersToSendBlock(t *testing.T) {
+	pm := &ProtocolManager{}
+	mockCtrl := gomock.NewController(t)
+
+	block := newBlock(123)
+	hash := block.Hash()
+
+	// 1. When the given node is a consensus node.
+	{
+		// len(pnsWithoutBlock) <= blockReceivingPNLimit
+		mockPeers := NewMockPeerSet(mockCtrl)
+		pm.nodetype = node.CONSENSUSNODE
+		pm.peers = mockPeers
+
+		mockCN := NewMockPeer(mockCtrl)
+		mockPN := NewMockPeer(mockCtrl)
+
+		mockPeers.EXPECT().CNWithoutBlock(hash).Return([]Peer{mockCN}).Times(1)
```

Comment:
Why use the time method? Is there a problem if not used?

Reply:
@nogang This is to make sure that the function works as expected. In this setting, I want `CNWithoutBlock(hash)` to be called exactly once and hence I injected the restriction.